In [45]:
import pandas as pd
import numpy as np
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.expand_frame_repr', False)

# Suppress all warnings
import warnings
warnings.filterwarnings("ignore")

In [46]:
df=pd.read_csv('all_matches.csv')

In [47]:
df.head(2)

match_id   season  start_date                  venue  innings  ball           batting_team                 bowling_team      striker  non_striker   bowler  runs_off_bat  extras  wides  noballs  byes  legbyes  penalty wicket_type player_dismissed  other_wicket_type  other_player_dismissed
0    335982  2007/08  2008-04-18  M Chinnaswamy Stadium        1   0.1  Kolkata Knight Riders  Royal Challengers Bangalore   SC Ganguly  BB McCullum  P Kumar             0       1    NaN      NaN   NaN      1.0      NaN         NaN              NaN                NaN                     NaN
1    335982  2007/08  2008-04-18  M Chinnaswamy Stadium        1   0.2  Kolkata Knight Riders  Royal Challengers Bangalore  BB McCullum   SC Ganguly  P Kumar             0       0    NaN      NaN   NaN      NaN      NaN         NaN              NaN                NaN                     NaN

In [48]:
df.columns

Index(['match_id', 'season', 'start_date', 'venue', 'innings', 'ball',
       'batting_team', 'bowling_team', 'striker', 'non_striker', 'bowler',
       'runs_off_bat', 'extras', 'wides', 'noballs', 'byes', 'legbyes',
       'penalty', 'wicket_type', 'player_dismissed', 'other_wicket_type',
       'other_player_dismissed'],
      dtype='object')

In [ ]:
df['striker'].unique()

array(['SC Ganguly', 'BB McCullum', 'RT Ponting', 'DJ Hussey',
       'Mohammad Hafeez', 'R Dravid', 'W Jaffer', 'V Kohli', 'JH Kallis',
       'CL White', 'MV Boucher', 'B Akhil', 'AA Noffke', 'P Kumar',
       'Z Khan', 'SB Joshi', 'PA Patel', 'ML Hayden', 'MEK Hussey',
       'MS Dhoni', 'SK Raina', 'JDP Oram', 'S Badrinath', 'K Goel',
       'JR Hopes', 'KC Sangakkara', 'Yuvraj Singh', 'SM Katich',
       'IK Pathan', 'T Kohli', 'YK Pathan', 'SR Watson', 'M Kaif',
       'DS Lehmann', 'RA Jadeja', 'M Rawat', 'D Salunkhe', 'SK Warne',
       'SK Trivedi', 'G Gambhir', 'V Sehwag', 'S Dhawan', 'AC Gilchrist',
       'Y Venugopal Rao', 'VVS Laxman', 'A Symonds', 'RG Sharma',
       'SB Styris', 'AS Yadav', 'SB Bangar', 'WPUJC Vaas', 'RP Singh',
       'WP Saha', 'LR Shukla', 'L Ronchi', 'ST Jayasuriya', 'DJ Thornely',
       'RV Uthappa', 'PR Shah', 'AM Nayar', 'SM Pollock',
       'Harbhajan Singh', 'S Chanderpaul', 'LRPL Taylor',
       'DPMD Jayawardene', 'S Sohal', 'B Lee', 'PP Cha

In [49]:
def phase(ball):


    if ball <= 5.6:
        return 'Powerplay'
    elif ball <= 14.6:
        return 'Middle'
    else:
        return 'Death'

In [50]:
df['phase'] = df['ball'].apply(lambda x: phase(x))

## Data where the striker is MS Dhoni

In [51]:
msd_df=df[(df['striker']=='MS Dhoni')]

## To check last Five years info. of Ms Dhoni

In [ ]:
df.dtypes

match_id                    int64
season                     object
start_date                 object
venue                      object
innings                   float64
ball                      float64
batting_team               object
bowling_team               object
striker                    object
non_striker                object
bowler                     object
runs_off_bat              float64
extras                    float64
wides                     float64
noballs                   float64
byes                      float64
legbyes                   float64
penalty                   float64
wicket_type                object
player_dismissed           object
other_wicket_type         float64
other_player_dismissed    float64
phase                      object
year                        int32
dtype: object

In [52]:
df['year']=pd.to_datetime(df['start_date']).dt.year


In [53]:
msd_df['year']=pd.to_datetime(msd_df['start_date']).dt.year

In [ ]:
msd_df.dtypes

match_id                    int64
season                     object
start_date                 object
venue                      object
innings                   float64
ball                      float64
batting_team               object
bowling_team               object
striker                    object
non_striker                object
bowler                     object
runs_off_bat              float64
extras                    float64
wides                     float64
noballs                   float64
byes                      float64
legbyes                   float64
penalty                   float64
wicket_type                object
player_dismissed           object
other_wicket_type         float64
other_player_dismissed    float64
phase                      object
year                        int32
dtype: object

In [54]:

# all players over the past five years Data
df_5y = df[((df['year'] == 2019) | (df['year'] == 2020) | (df['year'] == 2021) | (df['year'] == 20220) | (df['year'] == 2023) | (df['year'] == 2024))]

In [55]:
msd_5y=msd_df[((msd_df['year']==2019) | (msd_df['year']==2020) |  (msd_df['year']==2021) | (msd_df['year']==2022 )|(msd_df['year']==2023) | (msd_df['year']==2024))]

In [56]:
msd_5y

match_id   season  start_date                                              venue  innings  ball         batting_team                 bowling_team   striker       non_striker             bowler  runs_off_bat  extras  wides  noballs  byes  legbyes  penalty wicket_type  player_dismissed  other_wicket_type  other_player_dismissed      phase  year
151604   1175360     2019  2019-03-26                               Arun Jaitley Stadium        2  10.3  Chennai Super Kings               Delhi Capitals  MS Dhoni         KM Jadhav           A Mishra             4       0    NaN      NaN   NaN      NaN      NaN         NaN               NaN                NaN                     NaN     Middle  2019
151605   1175360     2019  2019-03-26                               Arun Jaitley Stadium        2  10.4  Chennai Super Kings               Delhi Capitals  MS Dhoni         KM Jadhav           A Mishra             0       1    NaN      NaN   NaN      1.0      NaN         NaN               NaN                NaN                     NaN     Middle  2019
151607   1175360     2019  2019-03-26                               Arun Jaitley Stadium        2  10.6  Chennai Super Kings               Delhi Capitals  MS Dhoni         KM Jadhav           A Mishra             1       0    NaN      NaN   NaN      NaN      NaN         NaN               NaN                NaN                     NaN     Middle  2019
151608   1175360     2019  2019-03-26                               Arun Jaitley Stadium        2  11.1  Chennai Super Kings               Delhi Capitals  MS Dhoni         KM Jadhav           KMA Paul             0       0    NaN      NaN   NaN      NaN      NaN         NaN               NaN                NaN                     NaN     Middle  2019
151609   1175360     2019  2019-03-26                               Arun Jaitley Stadium        2  11.2  Chennai Super Kings               Delhi Capitals  MS Dhoni         KM Jadhav           KMA Paul             1       0    NaN      NaN   NaN      NaN      NaN         NaN               NaN                NaN                     NaN     Middle  2019
151612   1175360     2019  2019-03-26                               Arun Jaitley Stadium        2  11.5  Chennai Super Kings               Delhi Capitals  MS Dhoni         KM Jadhav           KMA Paul             0       0    NaN      NaN   NaN      NaN      NaN         NaN               NaN                NaN                     NaN     Middle  2019
151613   1175360     2019  2019-03-26                               Arun Jaitley Stadium        2  11.6  Chennai Super Kings               Delhi Capitals  MS Dhoni         KM Jadhav           KMA Paul             1       0    NaN      NaN   NaN      NaN      NaN         NaN               NaN                NaN                     NaN     Middle  2019
151614   1175360     2019  2019-03-26                               Arun Jaitley Stadium        2  12.1  Chennai Super Kings               Delhi Capitals  MS Dhoni         KM Jadhav           AR Patel             2       0    NaN      NaN   NaN      NaN      NaN         NaN               NaN                NaN                     NaN     Middle  2019
151615   1175360     2019  2019-03-26                               Arun Jaitley Stadium        2  12.2  Chennai Super Kings               Delhi Capitals  MS Dhoni         KM Jadhav           AR Patel             1       0    NaN      NaN   NaN      NaN      NaN         NaN               NaN                NaN                     NaN     Middle  2019
151618   1175360     2019  2019-03-26                               Arun Jaitley Stadium        2  12.5  Chennai Super Kings               Delhi Capitals  MS Dhoni         KM Jadhav           AR Patel             0       0    NaN      NaN   NaN      NaN      NaN         NaN               NaN                NaN                     NaN     Middle  2019
151619   1175360     2019  2019-03-26                               Arun Jaitley Stadium        2  12.6  Chennai Super Kings 

In [ ]:
msd_5y['phase'].unique()

array([], dtype=object)

In [57]:
def balls_per_dismissal(balls, dismissals):
    if dismissals > 0:
        return balls / dismissals
    else:
        return balls / 1

def balls_per_boundary(balls, boundaries):
    if boundaries > 0:
        return balls / boundaries
    else:
        return balls / 1

def phasesOfplay(df, current_phase):

    df = df[df.phase == current_phase]
    df.reset_index(inplace=True, drop=True)

    df['isDot'] = df['runs_off_bat'].apply(lambda x: 1 if x == 0 else 0)
    df['isOne'] = df['runs_off_bat'].apply(lambda x: 1 if x == 1 else 0)
    df['isTwo'] = df['runs_off_bat'].apply(lambda x: 1 if x == 2 else 0)
    df['isThree'] = df['runs_off_bat'].apply(lambda x: 1 if x == 3 else 0)
    df['isFour'] = df['runs_off_bat'].apply(lambda x: 1 if x == 4 else 0)
    df['isSix'] = df['runs_off_bat'].apply(lambda x: 1 if x == 6 else 0)

    runs = pd.DataFrame(df.groupby(['striker', 'match_id'])['runs_off_bat'].sum().reset_index()).groupby(['striker'])['runs_off_bat'].sum().reset_index().rename(columns={'runs_off_bat': 'runs'})
    innings = pd.DataFrame(df.groupby(['striker'])['match_id'].apply(lambda x: len(list(np.unique(x)))).reset_index()).rename(columns={'match_id': 'innings'})
    balls = pd.DataFrame(df.groupby(['striker'])['match_id'].count()).reset_index().rename(columns={'match_id': 'balls'})
    dismissals = pd.DataFrame(df.groupby(['striker'])['player_dismissed'].count()).reset_index().rename(columns={'player_dismissed': 'dismissals'})

    dots = pd.DataFrame(df.groupby(['striker'])['isDot'].sum()).reset_index().rename(columns={'isDot': 'dots'})
    ones = pd.DataFrame(df.groupby(['striker'])['isOne'].sum()).reset_index().rename(columns={'isOne': 'ones'})
    twos = pd.DataFrame(df.groupby(['striker'])['isTwo'].sum()).reset_index().rename(columns={'isTwo': 'twos'})
    threes = pd.DataFrame(df.groupby(['striker'])['isThree'].sum()).reset_index().rename(columns={'isThree': 'threes'})
    fours = pd.DataFrame(df.groupby(['striker'])['isFour'].sum()).reset_index().rename(columns={'isFour': 'fours'})
    sixes = pd.DataFrame(df.groupby(['striker'])['isSix'].sum()).reset_index().rename(columns={'isSix': 'sixes'})

    df = pd.merge(innings, runs, on='striker').merge(balls, on='striker').merge(dismissals, on='striker').merge(dots, on='striker').merge(ones, on='striker').merge(twos, on='striker').merge(threes, on='striker').merge(fours, on='striker').merge(sixes, on='striker')

    # Strike Rate
    df['SR'] = 100 * (df['runs'] / df['balls'])

    # Runs per innings
    df['RPI'] = df['runs'] / df['innings']

    # Balls per dismissal
    df['BPD'] = df.apply(lambda x: balls_per_dismissal(x['balls'], x['dismissals']), axis=1)

    # Balls per boundary
    df['BPB'] = df.apply(lambda x: balls_per_boundary(x['balls'], (x['fours'] + x['sixes'])), axis=1)

    # Boundary Percentage
    df['Boundary %'] = (df['fours'] + df['sixes']) / df['balls'] * 100

    return df


In [58]:
# player's performance in Death Over
phasesOfplay(msd_5y, 'Death')


striker  innings  runs  balls  dismissals  dots  ones  twos  threes  fours  sixes          SR        RPI        BPD       BPB  Boundary %
0  MS Dhoni       58   855    536          28   192   176    49       1     65     53  159.514925  14.741379  19.142857  4.542373   22.014925

--------------------------------------------------------------------------------------------------------------------------------------------------------------

In [59]:
df_5y.head(3)

match_id season  start_date                   venue  innings  ball                 batting_team         bowling_team   striker non_striker     bowler  runs_off_bat  extras  wides  noballs  byes  legbyes  penalty wicket_type player_dismissed  other_wicket_type  other_player_dismissed      phase  year
150460   1175356   2019  2019-03-23  MA Chidambaram Stadium        1   0.1  Royal Challengers Bangalore  Chennai Super Kings   V Kohli    PA Patel  DL Chahar             1       0    NaN      NaN   NaN      NaN      NaN         NaN              NaN                NaN                     NaN  Powerplay  2019
150461   1175356   2019  2019-03-23  MA Chidambaram Stadium        1   0.2  Royal Challengers Bangalore  Chennai Super Kings  PA Patel     V Kohli  DL Chahar             0       0    NaN      NaN   NaN      NaN      NaN         NaN              NaN                NaN                     NaN  Powerplay  2019
150462   1175356   2019  2019-03-23  MA Chidambaram Stadium        1   0.3  Royal Challengers Bangalore  Chennai Super Kings  PA Patel     V Kohli  DL Chahar             0       0    NaN      NaN   NaN      NaN      NaN         NaN              NaN                NaN                     NaN  Powerplay  2019

In [60]:
df1=phasesOfplay(df_5y, 'Death')

In [61]:
df2=df1[df1['innings']>20]

In [62]:
df2.head(2)

striker  innings  runs  balls  dismissals  dots  ones  twos  threes  fours  sixes          SR        RPI        BPD       BPB  Boundary %
5  AB de Villiers       23   610    274          12    61    93    30       1     40     49  222.627737  26.521739  22.833333  3.078652   32.481752
7      AD Russell       36   734    399          27   170   100    11       0     48     70  183.959900  20.388889  14.777778  3.381356   29.573935

In [63]:

df2.sort_values(by='SR', ascending=False).reset_index()
df2 = df2.sort_values(by='SR', ascending=False).reset_index(drop=True)


In [64]:
df2.head(10)

striker  innings  runs  balls  dismissals  dots  ones  twos  threes  fours  sixes          SR        RPI        BPD       BPB  Boundary %
0  AB de Villiers       23   610    274          12    61    93    30       1     40     49  222.627737  26.521739  22.833333  3.078652   32.481752
1      AD Russell       36   734    399          27   170   100    11       0     48     70  183.959900  20.388889  14.777778  3.381356   29.573935
2         RR Pant       36   594    323          25    96   113    24       1     52     37  183.900929  16.500000  12.920000  3.629213   27.554180
3      KA Pollard       29   558    317          14   109   104    24       2     34     44  176.025237  19.241379  22.642857  4.064103   24.605678
4       SM Curran       28   384    220          17    70    72    21       0     36     21  174.545455  13.714286  12.941176  3.859649   25.909091
5       HH Pandya       46   751    433          30   142   149    38       0     49     55  173.441109  16.326087  14.433333  4.163462   24.018476
6        RK Singh       21   369    215          12    75    71    17       0     24     28  171.627907  17.571429  17.916667  4.134615   24.186047
7      KD Karthik       41   684    399          25   124   142    32       0     64     37  171.428571  16.682927  15.960000  3.950495   25.313283
8      SO Hetmyer       36   526    312          18    99   110    35       0     31     37  168.589744  14.611111  17.333333  4.588235   21.794872
9      MP Stoinis       34   499    296          19    92   105    27       0     46     26  168.581081  14.676471  15.578947  4.111111   24.324324

In [65]:
wt_sr, wt_rpi, wt_bpd, wt_bpb = 0.45 ,0.13,0.16,0.27


### Calculation - Normalizing all values to same dimensions

In [66]:
#step1: square of all values

df2['calc_SR'] = df2['SR'].apply(lambda x: x*x)
df2['calc_RPI'] = df2['RPI'].apply(lambda x: x*x)
df2['calc_BPD'] = df2['BPD'].apply(lambda x: x*x)
df2['calc_BPB'] = df2['BPB'].apply(lambda x: x*x)

sq_sr, sq_rpi, sq_bpd, sq_bpb = np.sqrt(df2[['calc_SR','calc_RPI', 'calc_BPD', 'calc_BPB']].sum(axis=0))

df2['calc_SR'] = df2['calc_SR'].apply(lambda x: x/sq_sr)
df2['calc_RPI'] = df2['calc_RPI'].apply(lambda x: x/sq_rpi)
df2['calc_BPD'] = df2['calc_BPD'].apply(lambda x: x/sq_bpd)
df2['calc_BPB'] = df2['calc_BPB'].apply(lambda x: x/sq_bpb)

df2['calc_SR'] = df2['calc_SR'].apply(lambda x: x*wt_sr)
df2['calc_RPI'] = df2['calc_RPI'].apply(lambda x: x*wt_rpi)
df2['calc_BPD'] = df2['calc_BPD'].apply(lambda x: x*wt_bpd)
df2['calc_BPB'] = df2['calc_BPB'].apply(lambda x: x*wt_bpb)

best_sr, worst_sr = max(df2['calc_SR']), min(df2['calc_SR'])
best_rpi, worst_rpi = max(df2['calc_RPI']), min(df2['calc_RPI'])
best_bpd, worst_bpd = max(df2['calc_BPD']), min(df2['calc_BPD'])
best_bpb, worst_bpb = min(df2['calc_BPB']), max(df2['calc_BPB'])


In [67]:
df2['dev_best_SR'] = df2['calc_SR'].apply(lambda x: (x-best_sr)*(x-best_sr))
df2['dev_best_RPI'] = df2['calc_RPI'].apply(lambda x: (x-best_rpi)*(x-best_rpi))
df2['dev_best_BPD'] = df2['calc_BPD'].apply(lambda x: (x-best_bpd)*(x-best_bpd))
df2['dev_best_BPB'] = df2['calc_BPB'].apply(lambda x: (x-best_bpb)*(x-best_bpb))

df2['dev_best_sqrt'] = df2.apply(lambda x: x['dev_best_SR'] + x['dev_best_RPI'] + x['dev_best_BPD'] + x['dev_best_BPB'], axis=1)

df2['dev_worst_SR'] = df2['calc_SR'].apply(lambda x: (x-worst_sr)*(x-worst_sr))
df2['dev_worst_RPI'] = df2['calc_RPI'].apply(lambda x: (x-worst_rpi)*(x-worst_rpi))
df2['dev_worst_BPD'] = df2['calc_BPD'].apply(lambda x: (x-worst_bpd)*(x-worst_bpd))
df2['dev_worst_BPB'] = df2['calc_BPB'].apply(lambda x: (x-worst_bpb)*(x-worst_bpb))

df2['dev_worst_sqrt'] = df2.apply(lambda x: x['dev_worst_SR'] + x['dev_worst_RPI'] + x['dev_worst_BPD'] + x['dev_worst_BPB'], axis=1)


In [68]:
df2['score'] = df2.apply(lambda x: x['dev_worst_sqrt']/(x['dev_worst_sqrt'] + x['dev_best_sqrt']), axis = 1)

In [69]:
df2[[ 'striker','innings', 'runs', 'balls', 'dismissals', 'BPB', 'score']].sort_values(['score'], ascending = False).reset_index(drop = True).head(25)

striker  innings  runs  balls  dismissals       BPB     score
0   AB de Villiers       23   610    274          12  3.078652  0.999874
1       AD Russell       36   734    399          27  3.381356  0.491192
2          RR Pant       36   594    323          25  3.629213  0.487154
3       KA Pollard       29   558    317          14  4.064103  0.320109
4        SM Curran       28   384    220          17  3.859649  0.284356
5        HH Pandya       46   751    433          30  4.163462  0.264845
6         RK Singh       21   369    215          12  4.134615  0.234551
7       KD Karthik       41   684    399          25  3.950495  0.230101
8       SO Hetmyer       36   526    312          18  4.588235  0.184937
9       MP Stoinis       34   499    296          19  4.111111  0.184325
10       V Shankar       24   319    191          16  4.897436  0.160888
11     Abdul Samad       26   371    225          18  4.500000  0.134087
12         R Parag       25   300    184          18  4.487805  0.112558
13        N Pooran       33   529    326          21  4.657143  0.105926
14        MS Dhoni       46   681    425          20  4.619565  0.088658
15        TH David       22   318    200          13  4.761905  0.075572
16          S Dube       25   372    237          17  4.836735  0.059972
17       RA Jadeja       45   697    447          18  5.079545  0.058529
18        KL Rahul       22   368    237          15  5.266667  0.049592
19       AT Rayudu       25   346    227          13  5.279070  0.034156
20       R Tewatia       34   447    297          18  4.950000  0.025602
21       DA Miller       26   343    228          15  4.750000  0.025119
22       KH Pandya       35   287    196          19  4.900000  0.011873
23     Rashid Khan       38   316    222          24  4.826087  0.004281
24        AR Patel       35   399    287          23  6.239130  0.001165

In [ ]:

#Group data by year and calculate total runs
yearly_runs = msd_df.groupby('year')['runs_off_bat'].sum().reset_index()

# Create Plotly bar plot
fig = go.Figure(data=[go.Bar(x=yearly_runs['year'], y=yearly_runs['runs_off_bat'])])

# Update layout
fig.update_layout(title='Dhoni\'s Year-wise Runs', xaxis_title='Year', yaxis_title='Runs')

# Show plot
fig.show()

In [71]:
last_df=df_5y[df_5y['ball']> 19.0]

In [73]:
df3=phasesOfplay(last_df, 'Death')

In [78]:
# top 10 Records :
df3[df3['innings']>10].sort_values(by='SR', ascending=False).reset_index(drop=True)[:10]


striker  innings  runs  balls  dismissals  dots  ones  twos  threes  fours  sixes          SR       RPI        BPD       BPB  Boundary %
0      MP Stoinis       15   124     44           4     9    10     2       0     14      9  281.818182  8.266667  11.000000  1.913043   52.272727
1       HH Pandya       17   152     56           6    19     8     4       0      7     18  271.428571  8.941176   9.333333  2.240000   44.642857
2  AB de Villiers       13   103     41           4     8    11     7       0      6      9  251.219512  7.923077  10.250000  2.733333   36.585366
3        MS Dhoni       27   247     99           6    28    21    12       0     13     25  249.494949  9.148148  16.500000  2.605263   38.383838
4      KA Pollard       18   149     60           6    15    12     9       1     11     12  248.333333  8.277778  10.000000  2.608696   38.333333
5       RA Jadeja       33   211     91          10    21    27    12       0     13     18  231.868132  6.393939   9.100000  2.935484   34.065934
6      AD Russell       15   104     51           8    20    12     1       0      9      9  203.921569  6.933333   6.375000  2.833333   35.294118
7       SM Curran       12    67     33           4     9     7     7       0      7      3  203.030303  5.583333   8.250000  3.300000   30.303030
8        RK Singh       12    81     40           4    18     9     0       0      3     10  202.500000  6.750000  10.000000  3.076923   32.500000
9        R Ashwin       12    56     28           4     8     8     4       0      4      4  200.000000  4.666667   7.000000  3.500000   28.571429

In [101]:
import plotly.graph_objs as go

# Filter top 10 records for the last over
top_10_last_over = df3[df3['innings'] > 10].sort_values(by='SR', ascending=False).head(10)

# Round strike rate to two decimal points
top_10_last_over['rounded_SR'] = top_10_last_over['SR'].round(2)

# Plot bar chart
fig = go.Figure(data=[go.Bar(x=top_10_last_over['striker'], y=top_10_last_over['SR'], marker_color='indianred', text=top_10_last_over['rounded_SR'], textposition='auto')])

# Update layout
fig.update_layout(
    title=dict(text='Top 10 Batsmans in the Last Over', font=dict(color='red', size=20)),
    xaxis_title=dict(text='Batsmans', font=dict(color='silver', size=16)),
    yaxis_title=dict(text='Strike Rate', font=dict(color='silver', size=16)),
    xaxis=dict(tickangle=45, tickfont=dict(color='red', size=12)),
    yaxis=dict(tickformat=".2f", tickfont=dict(color='orange', size=12)),
    plot_bgcolor='rgba(0,0,0,0)',
    paper_bgcolor='rgba(0,0,0,0)',
)

# Show the plot
fig.show()




#**Observation:**

When it comes to the final over in ipl matches over the last five years, MS Dhoni emerges as the 4th top performer based on his strike rate **249** . His ability to score quickly and effectively in these pressure-cooker situations highlights his exceptional skill and composure. It's a testament to his impact on the game, showcasing why he's regarded as one of the best in the business.

-----------------------------------------------------------------------------------------------------------------------------------------------

In [95]:
import plotly.graph_objs as go

# Filter top 10 records for the last over
top_scorers_last_20th_over = df3[df3['innings'] > 10].sort_values(by='runs', ascending=False).head(10)

# Plot bar chart
fig = go.Figure(data=[go.Bar(x=top_scorers_last_20th_over['striker'], y=top_scorers_last_20th_over['runs'], marker_color='indianred', text=top_scorers_last_20th_over['runs'], textposition='auto')])

# Update layout
fig.update_layout(
    title=dict(text='Top Scorers in the 20th Over (Last Five Years)', font=dict(color='red', size=20)),
    xaxis_title=dict(text='Batsmans', font=dict(color='silver', size=16)),
    yaxis_title=dict(text='Runs Scored', font=dict(color='silver', size=16)),
    xaxis=dict(tickangle=45, tickfont=dict(color='red', size=12)),
    yaxis=dict(tickformat=".2f", tickfont=dict(color='orange', size=12)),
    plot_bgcolor='rgba(0,0,0,0)',
    paper_bgcolor='rgba(0,0,0,0)',
    font=dict(color='black', size=12),
)

# Show the plot
fig.show()


#**Observation:**
In the bar graph depicting the top 10 batsmen in the last over, it's notable that MS Dhoni stands out with the highest number of runs, registering a significant 247 runs. This observation underscores Dhoni's prowess and effectiveness as a finisher in the final overs of cricket matches. His ability to consistently perform and accumulate runs in the concluding stages of the game highlights his strategic importance and skill in steering his team towards victory. This statistical insight not only emphasizes Dhoni's exceptional batting capabilities but also underscores the pivotal role he plays in shaping the outcome of matches during critical junctures.

------------------------------------------------------------------------------------------------------------------------------------------------


#Based on our analysis, MS Dhoni continues to display impressive batting skills, highlighting his importance to the Chennai Super Kings (CSK) . His knack for maintaining a quick scoring rate and adding valuable runs emphasizes his lasting impact in today's cricket. Therefore, it's clear that Dhoni remains a strong batsman, capable of making big contributions to CSK's success .


----------------------------------------------------------------------